### PR TITLE
Combined PR that addresses #88 and #89 and #148 in one

### PR DESCRIPTION
**Pull request by onovotny**
I've merged the two into a single PR as I'm using it in an unofficial build for now....just passing it along.

### CODE REVIEW

File: Refit/RequestBuilderImplementation.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 namespace Refit
 {
-    public class RequestBuilderFactory : IRequestBuilderFactory
+    internal class RequestBuilderFactory : IRequestBuilderFactory
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: ~~internal~~

File: Refit/RequestBuilderImplementation.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ public IRequestBuilder Create(Type interfaceType, RefitSettings settings = null)
         }
     }
 
-    public class RequestBuilderImplementation : IRequestBuilder
+    internal class RequestBuilderImplementation : IRequestBuilder
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: ~~internal~~

File: Refit/RequestBuilderImplementation.cs [hunk 3]
```diff
@@ -52,7 +52,7 @@ public RequestBuilderImplementation(Type targetInterface, RefitSettings refitSet
             get { return interfaceHttpMethods.Keys; }
         }
 
-        public Func<object[], HttpRequestMessage> BuildRequestFactoryForMethod(string methodName, string basePath = "")
+        private Func<object[], HttpRequestMessage> buildRequestFactoryForMethod(string methodName, string basePath, bool paramsContainsCancellationToken)
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: ~~private~~

File: Refit/RequestBuilderImplementation.cs [hunk 4]
```diff
@@ -60,6 +60,12 @@ public RequestBuilderImplementation(Type targetInterface, RefitSettings refitSet
             var restMethod = interfaceHttpMethods[methodName];
 
             return paramList => {
+                // make sure we strip out any cancelation tokens
+                if (paramsContainsCancellationToken)
+                {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 5]
```diff
@@ -126,7 +132,7 @@ public RequestBuilderImplementation(Type targetInterface, RefitSettings refitSet
                     // for anything that fell through to here, if this is not
                     // a multipart method, add the parameter to the query string
                     if (!restMethod.IsMultipart) {
-                        queryParamsToAdd[restMethod.QueryParameterMap[i]] = settings.UrlParameterFormatter.Format(paramList[i], restMethod.ParameterInfoMap[i]);
+                            queryParamsToAdd[restMethod.QueryParameterMap[i]] = settings.UrlParameterFormatter.Format(paramList[i], restMethod.ParameterInfoMap[i]);
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: Extra tab here...

File: Refit/RequestBuilderImplementation.cs [hunk 6]
```diff
@@ -272,12 +278,21 @@ void addMultipartItem(MultipartFormDataContent multiPartContent, string itemName
...
             return async (client, paramList) => {
-                var factory = BuildRequestFactoryForMethod(restMethod.Name, client.BaseAddress.AbsolutePath);
+                var factory = buildRequestFactoryForMethod(restMethod.Name, client.BaseAddress.AbsolutePath, restMethod.CancellationToken != null);
                 var rq = factory(paramList);
-                var resp = await client.SendAsync(rq);
+
+                var ct = CancellationToken.None;
+
+                if (restMethod.CancellationToken != null)
+                {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 7]
```diff
@@ -536,12 +583,33 @@ string getAttachmentNameForParameter(ParameterInfo paramInfo)
                 throw new ArgumentException("Only one parameter can be a Body parameter");
             }
 
-            if (bodyParams.Count == 0) {
-                return null;
+            // #1, body attribute wins
+            if (bodyParams.Count == 1)
+            {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 8]
```diff
@@ -536,12 +583,33 @@ string getAttachmentNameForParameter(ParameterInfo paramInfo)
...
+                var ret = bodyParams[0];
+                return Tuple.Create(ret.BodyAttribute.SerializationMethod, parameterList.IndexOf(ret.Parameter));
+            }
+
+            // No body attribute found, check rule #2
+         
+            // see if we're a post/put/patch
+            if (method.Equals(HttpMethod.Post) || method.Equals(HttpMethod.Put) || method.Equals(patchMethod))
+            {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 9]
```diff
@@ -536,12 +583,33 @@ string getAttachmentNameForParameter(ParameterInfo paramInfo)
...
+
+                var refParams = parameterList.Where(pi => !pi.ParameterType.GetTypeInfo().IsValueType && pi.ParameterType != typeof(string)).ToList();
+                
+                // Check for rule #3
+                if (refParams.Count > 1)
+                {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 10]
```diff
@@ -536,12 +583,33 @@ string getAttachmentNameForParameter(ParameterInfo paramInfo)
...
+                    throw new ArgumentException("Multiple complex types found. Specify one parameter as the body using BodyAttribute");
+                }
+                else if (refParams.Count == 1)
+                {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 11]
```diff
@@ -536,12 +583,33 @@ string getAttachmentNameForParameter(ParameterInfo paramInfo)
...
+            {
+
+                var refParams = parameterList.Where(pi => !pi.ParameterType.GetTypeInfo().IsValueType && pi.ParameterType != typeof(string)).ToList();
+                
+                // Check for rule #3
+                if (refParams.Count > 1)
+                {
+                    throw new ArgumentException("Multiple complex types found. Specify one parameter as the body using BodyAttribute");
+                }
+                else if (refParams.Count == 1)
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

**Comment by bennor** [comment]
This all looks great other than the cosmetic stuff. :+1:

File: Refit/RequestBuilderImplementation.cs [hunk 12]
```diff
@@ -285,16 +300,21 @@ void addMultipartItem(MultipartFormDataContent multiPartContent, string itemName
         Func<HttpClient, object[], Task<T>> buildTaskFuncForMethod<T>(RestMethodInfo restMethod)
         {
             var ret = buildCancellableTaskFuncForMethod<T>(restMethod);
-            return (client, paramList) => ret(client, CancellationToken.None, paramList);
+            return (client, paramList) =>
+                   {
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: :arrow_up:

File: Refit/RequestBuilderImplementation.cs [hunk 13]
```diff
@@ -330,7 +351,18 @@ void addMultipartItem(MultipartFormDataContent multiPartContent, string itemName
             var taskFunc = buildCancellableTaskFuncForMethod<T>(restMethod);
 
             return (client, paramList) => 
-                new TaskToObservable<T>(ct => taskFunc(client, ct, paramList));
+                new TaskToObservable<T>(ct =>
+                                        {
+                                            var methodCt = CancellationToken.None;
```
Comment (bennor) [inline Refit/RequestBuilderImplementation.cs]: Not sure exactly how @paulcbetts would want this formatted, but this looks like R# formatting. :smiling_imp:

**Comment by onovotny** [comment]
Thanks, fixed the formatting to match

**Comment by mburbea** [comment]
Is this going to be merged? This branch looks to solve a lot of my annoyance with this package.

**Pull request merged**